### PR TITLE
feat(eslint-plugin): [no-unsafe-access] nounsafeAccess

### PR DESCRIPTION
**Pull request by jiameng123**
Feat(eslint-plugin): "Non-built-in member access enforces the use of optional chains"

### CODE REVIEW

**Review by bradzacher — CHANGES_REQUESTED** [review]
Thanks for the PR.

However - this rule is not something that we'll accept into this plugin.

What it enforces completely misses the point of TypeScript and type definitions.
There is nothing unsafe about the following code

```ts
interface IAjaxData {
  error: boolean;
  errorCode: string;
  data: {
    anyB: {
      anyC: string;
    };
  };
}
const data: IAjaxData = await fetchData('url');
const c = data.data.anyB.anyC;
```

This code is, as defined, completely safe.

So converting the final line to `data?.data?.anyB?.anyC` is unnecessary and ultimately wasteful.
It's wasteful because you're doing unnecessary checks now. `?.` is not a free operator - so doing it when not needed is not good practice.

If you define your types correctly using the optional property annotation:
```ts
interface IAjaxData {
  error: boolean;
  errorCode: string;
  data?: {
    anyB?: {
      anyC?: string;
    };
  };
}
const data: IAjaxData = await fetchData('url');
const c = data.data.anyB.anyC;
```

Now all of a sudden this code is *unsafe*, because the properties are declared to be optional. Thankfully TS knows this and understands this! TS will now error on our code:
image

So now we know we *have* to use `?.`, and we can add them in to make the code safe.

Another reason that this practice is bad is because TypeScript will not give you correct types when you unnecessarily use `?.`.

For the following code:
```ts
interface Foo { a: string };
declare const foo: Foo;

const one = foo.a; // TS type === string
const two = foo?.a; // TS type === string !!!!!!
```

Note that the second example, the type ***IS NOT*** `string | undefined`. Why? Because TS assumes the types are correct, and that the `?.` does nothing.

So unnecessarily annotating `?.` gives you a ***false sense of security***!

---

To summarise:
Converting every single member expression to use `?.` is an actively *bad* practice.
You should instead ensure your types are *correct*, and when you are converting across an unsafe boundary (eg parsing JSON data), you should ensure your types are declared with optional properties for safety.

**Comment by jiameng123** [comment]
@bradzacher  thanks for the thorough explanation. This PR primarily wants to resolve the error when the runtime ajax returns data as null (which can cause a program to crash if the developer forgets to do the error handling). Force all access to '?.' There's a real waste, and perhaps using a rule like ````Promise< (data::a: number)> to cast to Promise< Partial <(data:a: number>>```` is a reminder to developers.  What do you think of the suggestion, please?

**Comment by bradzacher** [comment]
I would consider building a layer on top of your fetch library which enables you to do this automatically.

For example

```
function myFetch<T extends object = unknown>(uri: string): DeepPartial<T> {
  return fetch(uri);
}
```

(where `DeepPartial` is a type which applies partial in a nested fashion to an object type)

That way you don't need to worry about a lint rule to enforce certain calls are typed in certain ways because your tooling will just do it automatically!

**Comment by jiameng123** [comment]
Thank you very much for taking the time to do this work
